api/services/locations.js

```diff
@@ -0,0 +1,170 @@
+const checkQueryStatusInterval = 2000;
+let intervalId = false;
+let QueryExecutionId;
+
+const checkQueryStatus = async function () {
+  function stopListener () {
+    intervalId = clearInterval(intervalId);
+  }
+
+  try {
+    const data = await athena.getQueryExecution({ QueryExecutionId }).promise();
+
+    // If response is malformed, stop listener.
+    if (!has(data, 'QueryExecution.Status.State')) {
+      console.log('Error while fetching Athena query status.');
+      stopListener();
+      return;
+    }
+
+    // Process state
+    const { State } = data.QueryExecution.Status;
+    if (State === 'SUCCEEDED') {
+      // update locations
+      console.log('Athena query was finished successfully!');
+      stopListener();
+      upsertLocations();
+    } else if (State !== 'RUNNING') {
+      console.log('Athena query was interrupted.');
+      stopListener();
+    } else {
+      console.log('Athena query is running...');
+    }
+  } catch (err) {
+    // Unexpected error occurred.
+    console.log(err, err.stack);
+    stopListener();
+  }
+};
+
+export const fetchLocations = async function () {
+  // Check if a query is already running
+  if (!intervalId) {
+    console.log('Creating Athena query.');
+    const data = await athena
+      .startQueryExecution({
+        QueryString,
+        QueryExecutionContext: {
+          Database
+        },
+        ResultConfiguration: {
+          OutputLocation
+        }
+      })
+      .promise();
+
+    // Register query run and start listener
+    QueryExecutionId = data.QueryExecutionId;
+    intervalId = setInterval(checkQueryStatus, checkQueryStatusInterval);
+  }
+};
+
+export const upsertLocations = async function () {
+  let queryResults = await athena
+    .getQueryResults({ QueryExecutionId })
+    .promise();
+
+  if (has(queryResults, 'ResultSet.Rows')) {
+    let rows = queryResults.ResultSet.Rows;
+
+    // Flatten "Data" and "VarCharValue" properties used by Athena
+    rows = rows.map(r => {
+      r = r.Data;
+      return r.map(value => value.VarCharValue);
+    });
+
+    // Get available properties
+    const properties = rows.shift();
+
+    // Parse rows into JSON
+    let locations = rows.map(r => {
+      const location = {};
+      for (let i = 0; i < r.length; i++) {
+        location[properties[i]] = r[i];
+        location.lon = parseFloat(location.lon);
+        location.lat = parseFloat(location.lat);
+      }
+      return location;
+    });
+
+    // Sort locations
+    locations = sortBy(locations, ['country', 'firstUpdated']);
+
+    // Get number of locations per country
+    let locationsPerCountry = await db
+      .select(db.raw('country, count(country)'))
+      .from('locations')
+      .groupBy('country');
+
+    // Transform results for easier access
+    locationsPerCountry = locationsPerCountry.reduce((acc, r) => {
+      acc[r.country] = parseInt(r.count);
+      return acc;
+    }, {});
+
+    for (const location of locations) {
```

Comment:
This is being done on locations right? We've currently got 10,353 so this would loop and make 10,353 database requests every time this function was run, right? That sounds bad, but I am willing to be told otherwise. Would it make more sense to pull the whole table into memory once and then compare against that?